### DISCUSSION TITLE
When route component gets rendered?

### DISCUSSION THREAD

**Discussion by Dragomir-Ivanov** [TanStack/router#1351] [General]
Greetings,
Trying to wrap my mind around TSR, and against my use-case. I am using search params, and it seems that the whole route component is re-rendered when search param get changes. This is not intuitive and causes a lot of head scratching.

Given how these params are accessed via `useSearch` hook, wouldn't it more logical this hook to trigger the rerender of possibly only needed component, instead of the whole root component. 

If the present behaviour is really intentional, is there any React recipe how to focus re-rendering only where it is needed, instead of the whole route component.

Thank you!

**Comment by TkDodo** [comment]
Can you show a minimal reproduction because afaik this shouldn't happen

**Reply by Dragomir-Ivanov** [reply]
Yes,
I have prepared a demo here:
https://stackblitz.com/edit/vitejs-vite-qfih43

I have updated all deps to their latest major versions, including TSR. Look at the `About` route, I have added a `<select>` that manipulates `searchParams`. On a change, `About` component also rerenders, even I don't consume the params via `useSearch`. 
I need to use `memo` in order for the `About` not to render.

Thank you for your fast reaction.

**Reply by TkDodo** [reply]
okay I misunderstood your statement:

> wouldn't it more logical this hook to trigger the rerender of possibly only needed component, instead of the whole root component.

the root route isn't re-rendering - the component for the route you are navigating to is. I guess it could be optimized, or maybe it's a bug, not sure. I think it's a good idea and quite easy to achieve to just wrap the `component` for your route in `React.memo`, as it doesn't receive any props:

```diff
const aboutRoute = createRoute({
   getParentRoute: () => rootRoute,
   path: '/about',
-  component: About,
+  component: React.memo(About),
```

I have seen some other situations where components re-render even though I think they shouldn't. Maybe the router could just memoize the component that we pass to `component` internally. Let's see what maintainers have to say. I'm personally fine with manual memoization here when needed.

**Reply by Dragomir-Ivanov** [reply]
Hi there,
Yes, I have prepared Memoed route in the demo, I thought you would see it. Used exactly your proposed solution.
However I found something else ( not saying it is a bug, but surely not documented or logical).

```
const memoedRoute = createRoute({
  getParentRoute: () => rootRoute,
  path: '/memoed',
  component: React.memo(Memoed),
  validateSearch: (search: Record<string, unknown>) => ({
    perPage: search.perPage,
    q: search.q,
  }),
  // loaderDeps: ({ search }) => [search.perPage, search.q],
});
```

This is the route for the mentioned `Memoed` component. If I type in the `<input>` with `loaderDeps` commented, the `MemoedInternals` component isn't re-rendered, however as soon as `loaderDeps` is not commented, the component is re-rendered on every input change. This means that `useNavigate` triggers re-render. Why is that? Granted, if I used `useLoaderDeps` or even `useLoaderData` yes, but `useNavigate`?

**Reply by Dragomir-Ivanov** [reply]
Thinking more about this, I think that 
```
createRoute({
...
componet: Component
})
```
Should be rendered only for the initial route match. For the subsequent route changes: route params, search params, sub-routes, this component, shouldn't re-render. If such re-render is needed, the user would have probably used the `useParams`, `useSearch`, etc. hooks, with appropriate `select` to narrow data of interest. 
That is my expectation as a React developer facing TSR for the first time. 

Using TSR with `uncontrolled` inputs as it is now, is a pain. Say I want an `<input defaultValue={initialSearchParam}>` to be uncontrolled and be initialized only with the initial search param. Other parts of the application can subscribe to `useSearch()` and consume the value of the `<input>`. Also `uncontrolled` input can be debounced for `naviate.()`, without affecting it's actual value.

**Reply by TkDodo** [reply]
seems like `useNavigate` subscribes to the whole `pathname`:

https://github.com/tanstack/router/blob/c2fbb56b251d10f2c2832916ea60a779c079ae24/packages/react-router/src/useNavigate.tsx#L27-L30

I think this can be optimized by imperatively accessing the `pathname` in the returned function rather than subscribing to it. Let me see if I can PR that :)

**Reply by Dragomir-Ivanov** [reply]
Hmm, yes whole `pathname` might be problematic, but why re-renders only when I specify `loaderDeps`? Isn't `pathname` independent from `loaderDeps` result?

**Reply by TkDodo** [reply]
https://github.com/TanStack/router/pull/1359

**Reply by Dragomir-Ivanov** [reply]
Thank you @TkDodo ! 

Can we discuss the second part of my question https://github.com/TanStack/router/discussions/1351#discussioncomment-8880271. 

I have updated the Demo above, the `Memoed` Component. 
It is basically `uncontrolled` `<input>`, that modifies a `searchParam`. There should be a way to get the initial search params, when the router opened a route, but without `subscribing` for further updates, i.e. without any re-renreders of the component.
So I can initialize the `<input>`'s `defaultValue` with. In other parts of the application components may subscribe to `useSearch` or `useLoaderData` in order to get needed information and react to changes of the `<input>`. 
I essence use URL search params as a basic state store.

**Reply by TkDodo** [reply]
I think you can do `useRouter().state`, which is imperative and doesn't subscribe you to anything. I'd want to rename it to `useRouter().getState()` to make that more obvious.

But honestly, if you just extract the `input` with the `defaultValue` to it's own component, the re-renders really shouldn't matter much. With the above solution, you run the risk in some other part of your code to also read those deps and expect them to be up-to-date. Just because your use-case right now only needs them for `defaultValue` doesn't mean it'll stay that way.

**Reply by Dragomir-Ivanov** [reply]
Hmm, `useRouter().state` would be perfect if I could extract `loaderDeps` and `loaderData` from it, and cache it on the first component render. However `state` doesn't expose these. Can we expose these?

When `<input>` is controlled and constantly re-rendered it becomes laggy. Granted in my use-case I could live with it, but it would be a good optimization, and TSR could cover wider use-cases. We want TSR to be the BEST!
Currently in my web apps, I am using `router5`, which is pretty much dead, but works.

**Reply by TkDodo** [reply]
everything is available on `state`, maybe not as easily explorable. `getRenderedMatches(state)` should give you matches, which will contain most of what you need.

**Reply by Dragomir-Ivanov** [reply]
Thank you @TkDodo. I dug through the sources, and found the function. This should work imperatively, and not reactive.
Will try asap, and comment/close the issue.

**Reply by Dragomir-Ivanov** [reply]
That's it then. It was `getRenderedMatches(state)` all I needed all along. I think this function needs more love in the Docs :)
I managed to optimize my rendering behavior for uncontrolled input elements.

Thanks @TkDodo. If I catch something else, surely I will report.

**Reply by TkDodo** [reply]
Please consider contributing to the docs :)

**Reply by Dragomir-Ivanov** [reply]
Will do in the coming days.
